src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala

```diff
@@ -0,0 +1,351 @@
+
+    case "waitingForInFlight" => sender() ! "waitingForInFlight" // for testing
+  }
+
+  private[this] def receiveStop: Receive = {
+    case AppTaskLauncherActor.Stop => waitingForInFlight()
+  }
+
+  private[this] def waitingForInFlight(): Unit = {
+    if (inFlightTaskLaunches.isEmpty) {
+      context.stop(self)
+    }
+    else {
+      context.become(receiveWaitingForInFlight)
+    }
+  }
+
+  /**
+    * Receive rate limiter updates.
+    */
+  private[this] def receiveDelayUpdate: Receive = {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      if (backOffUntil.forall(_ < delayUntil)) {
+        import context.dispatcher
+        val now: Timestamp = clock.now()
+        recheckBackOff.foreach(_.cancel())
+        recheckBackOff = Some(
+          context.system.scheduler.scheduleOnce(now until delayUntil, self, RecheckIfBackOffUntilReached)
+        )
+      }
+
+      backOffUntil = Some(delayUntil)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+      log.debug("After delay update {}", status)
+
+    case RecheckIfBackOffUntilReached => OfferMatcherRegistration.manageOfferMatcherStatus()
+  }
+
+  private[this] def receiveTaskLaunchNotification: Receive = {
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) if inFlightTaskLaunches(taskInfo.getTaskId) =>
+      // This task is not yet known to mesos, so there will be no event that removes
+      // it automatically from the taskTracker.
+      taskTracker.terminated(app.id, taskInfo.getTaskId.getValue)
+      removeTask(taskInfo.getTaskId)
+      tasksToLaunch += 1
+      log.info(
+        "Task launch for '{}' was denied, reason '{}', rescheduling. {}",
+        taskInfo.getTaskId.getValue, reason, status)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) => // ignore
+
+    case TaskLaunchSourceDelegate.TaskLaunchAccepted(taskInfo) =>
+      inFlightTaskLaunches -= taskInfo.getTaskId
+      log.info("Task launch for '{}' was accepted. {}", taskInfo.getTaskId.getValue, status)
+  }
+
+  private[this] def receiveTaskStatusUpdate: Receive = {
+    case TaskStatusUpdate(_, taskId, MarathonTaskStatus.Terminal(_)) =>
+      log.debug("task '{}' finished", taskId.getValue)
+      removeTask(taskId)
+
+    case TaskStatusUpdate(_, taskId, _) => log.debug("ignore update of task '{}'", taskId.getValue)
+  }
+
+  private[this] def removeTask(taskId: TaskID): Unit = {
+    inFlightTaskLaunches -= taskId
+    runningTasksMap.get(taskId.getValue).foreach { marathonTask =>
+      runningTasksMap -= taskId.getValue
+      runningTasks -= marathonTask
+    }
+  }
+
+  private[this] def receiveGetCurrentCount: Receive = {
+    case AppTaskLauncherActor.GetCount =>
+      replyWithQueuedTaskCount()
+  }
+
+  private[this] def receiveAddCount: Receive = {
+    case AppTaskLauncherActor.AddTasks(newApp, addCount) =>
+      if (app != newApp) {
```

Comment:
If the app definition is updated. We want to set the tasks to launch to addCount in this case, though.